### PR TITLE
[libc] implement mlock/mlock2/munlock/mlockall/munlockall

### LINKED ISSUE #79336: [libc] implement mlock/munlock
https://github.com/llvm/llvm-project/pull/79309 is where it would be nice to use these.

cc @SchrodingerZhu

### PR DESCRIPTION
**Pull request by SchrodingerZhu**
fixes #79336

### CODE REVIEW

**Review by nickdesaulniers — APPROVED** [review]

**Comment by SchrodingerZhu** [comment]
@nickdesaulniers putting `mlock2` inside `mlock` does not work as the header would not be generated properly. I will separate them.

**Review by nickdesaulniers — APPROVED** [review]

File: libc/test/src/sys/mman/linux/mlock_test.cpp [hunk 1]
```diff
@@ -49,6 +51,28 @@ struct PageHolder {
...
 };
 
+static bool get_capacity(unsigned int cap) {
+  struct __user_cap_header_struct header;
+  header.pid = 0;
+  header.version = _LINUX_CAPABILITY_VERSION_3;
+  struct __user_cap_data_struct data[_LINUX_CAPABILITY_U32S_3];
+  long res = LIBC_NAMESPACE::syscall_impl(
+      SYS_capget, LIBC_NAMESPACE::cpp::bit_cast<long>(&header),
+      LIBC_NAMESPACE::cpp::bit_cast<long>(&data));
```
Comment (nickdesaulniers) [inline libc/test/src/sys/mman/linux/mlock_test.cpp:61]: File a TODO to implement this next. :smiley:

File: libc/test/src/sys/mman/linux/mlock_test.cpp [hunk 2]
```diff
@@ -49,6 +51,28 @@ struct PageHolder {
   bool is_valid() { return addr != MAP_FAILED; }
 };
 
+static bool get_capacity(unsigned int cap) {
+  struct __user_cap_header_struct header;
```
Comment (nickdesaulniers) [inline libc/test/src/sys/mman/linux/mlock_test.cpp:55]: The `struct` keyword is not necessary in C++. Please remove.

File: libc/test/src/sys/mman/linux/mlock_test.cpp [hunk 3]
```diff
@@ -49,6 +51,28 @@ struct PageHolder {
...
+  header.pid = 0;
+  header.version = _LINUX_CAPABILITY_VERSION_3;
+  struct __user_cap_data_struct data[_LINUX_CAPABILITY_U32S_3];
+  long res = LIBC_NAMESPACE::syscall_impl(
+      SYS_capget, LIBC_NAMESPACE::cpp::bit_cast<long>(&header),
+      LIBC_NAMESPACE::cpp::bit_cast<long>(&data));
+  if (res < 0)
+    return false;
+  auto idx = CAP_TO_INDEX(cap);
+  auto shift = CAP_TO_MASK(cap);
```
Comment (nickdesaulniers) [inline libc/test/src/sys/mman/linux/mlock_test.cpp:65]: Please don't use `auto` here.

File: libc/test/src/sys/mman/linux/mlock_test.cpp [hunk 4]
```diff
@@ -49,6 +51,28 @@ struct PageHolder {
...
+  return (data[idx].effective & shift) != 0;
+}
+
+static bool is_permitted_size(size_t size) {
+  struct rlimit rlimits;
```
Comment (nickdesaulniers) [inline libc/test/src/sys/mman/linux/mlock_test.cpp:70]: remove `struct`

File: libc/test/src/sys/mman/linux/mlock_test.cpp [hunk 5]
```diff
@@ -57,9 +81,12 @@ TEST(LlvmLibcMlockTest, UnMappedMemory) {
 TEST(LlvmLibcMlockTest, Overflow) {
   PageHolder holder;
   EXPECT_TRUE(holder.is_valid());
-  EXPECT_THAT(LIBC_NAMESPACE::mlock(holder.addr, -holder.size), Fails(EINVAL));
-  EXPECT_THAT(LIBC_NAMESPACE::munlock(holder.addr, -holder.size),
-              Fails(EINVAL));
+  size_t negative_size = -holder.size;
+  auto expected_errno = is_permitted_size(negative_size) ? EINVAL : ENOMEM;
```
Comment (nickdesaulniers) [inline libc/test/src/sys/mman/linux/mlock_test.cpp:85]: no `auto`

Comment (nickdesaulniers) [reply libc/test/src/sys/mman/linux/mlock_test.cpp:85]: https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable

File: libc/test/src/sys/mman/linux/mlock_test.cpp [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+  char &operator[](size_t i) { return reinterpret_cast<char *>(addr)[i]; }
+
+  bool is_valid() { return addr != MAP_FAILED; }
+};
+
+static bool get_capacity(unsigned int cap) {
+  struct __user_cap_header_struct header;
+  header.pid = 0;
+  header.version = _LINUX_CAPABILITY_VERSION_3;
+  struct __user_cap_data_struct data[_LINUX_CAPABILITY_U32S_3];
```
Comment (nickdesaulniers) [inline libc/test/src/sys/mman/linux/mlock_test.cpp:58]: no struct

**Review by nickdesaulniers — APPROVED** [review]

**Comment by SchrodingerZhu** [comment]
emmm, is this a clang format bug?

**Comment by SchrodingerZhu** [comment]
I will merge this since the absorbed test is relatively simple.

File: libc/test/src/sys/mman/linux/mlock_test.cpp [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+                                    MAP_ANONYMOUS | MAP_PRIVATE, -1, 0);
+  EXPECT_EQ(0, libc_errno);
+  EXPECT_NE(addr, MAP_FAILED);
+
+  // Invalid mlock2 flags.
+  EXPECT_THAT(LIBC_NAMESPACE::mlock2(addr, alloc_size, 1234), Fails(EINVAL));
+
+  // Invalid mlockall flags.
+  EXPECT_THAT(LIBC_NAMESPACE::mlockall(1234), Fails(EINVAL));
+  EXPECT_THAT(LIBC_NAMESPACE::mlockall(MCL_ONFAULT), Fails(EINVAL));
```
Comment (nickdesaulniers) [inline libc/test/src/sys/mman/linux/mlock_test.cpp:134]: This is failing in post submit. PTAL

Comment (nickdesaulniers) [reply libc/test/src/sys/mman/linux/mlock_test.cpp:134]: https://github.com/llvm/llvm-project/pull/80075

**Pull request merged**